### PR TITLE
Accept Ubuntu Cosmic's Python version, too.

### PR DESCRIPTION
**Pull request by svenpanne**
Without this workaround pythonfinder doesn't crash, but it simply ignores
Ubuntu Cosmic's "2.7.15+" version.  Alas, that version does not conform to
PEP 440, but there is nothing one can do about that.  Nevertheless,
pythonfinder (and consequently pipenv!) should better accept that version,
anyway.  Adding that workaround in pythonfinder seems to be the right place,
making the VERSION_PATTERN regex in packaging's version.py more forgiving
doesn't look right.

### CODE REVIEW

**Comment by techalchemy** [comment]
Hmmm I thought that regex worked, I tested it!

**Comment by techalchemy** [comment]
We need a way to greedily consume numbers or dots until we hit a different character.

**Comment by svenpanne** [comment]
> We need a way to greedily consume numbers or dots until we hit a different character.

Do you mean: Replace both `if version.endswith(...):` statements by "keep only the prefix consisting of numbers and dots"?

**Comment by techalchemy** [comment]
Can you show me what happens on `2.7.15+` with this version?

**Comment by techalchemy** [comment]
That is, with the master branch

**Comment by techalchemy** [comment]
can you do me a favor and try the branch that is up on the pr over at #46:

`pip install -e git+https://github.com/sarugaku/pythonfinder.git@bugfix/parse-versions#egg=pythonfinder`

**Comment by svenpanne** [comment]
> can you do me a favor and try the branch that is up on the pr over at #46:
> 
> `pip install -e git+https://github.com/sarugaku/pythonfinder.git@bugfix/parse-versions#egg=pythonfinder`

Alas, the branch doesn't work, probably a missing dependency for `typing`. Running `pyfinder --findall` gives a traceback:

```
Traceback (most recent call last):
  File "/home/sven/x/.venv/bin/pyfinder", line 11, in <module>
    load_entry_point('pythonfinder', 'console_scripts', 'pyfinder')()
  File "/home/sven/x/.venv/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 487, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/home/sven/x/.venv/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2728, in load_entry_point
    return ep.load()
  File "/home/sven/x/.venv/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2346, in load
    return self.resolve()
  File "/home/sven/x/.venv/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2352, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/home/sven/x/.venv/src/pythonfinder/src/pythonfinder/__init__.py", line 14, in <module>
    from .pythonfinder import Finder
  File "/home/sven/x/.venv/src/pythonfinder/src/pythonfinder/pythonfinder.py", line 6, in <module>
    from .models import SystemPath
  File "/home/sven/x/.venv/src/pythonfinder/src/pythonfinder/models/__init__.py", line 11, in <module>
    from ..utils import KNOWN_EXTS, unnest
  File "/home/sven/x/.venv/src/pythonfinder/src/pythonfinder/utils.py", line 18, in <module>
    from .environment import PYENV_ROOT, ASDF_DATA_DIR, MYPY_RUNNING
  File "/home/sven/x/.venv/src/pythonfinder/src/pythonfinder/environment.py", line 33, in <module>
    MYPY_RUNNING = os.environ.get("MYPY_RUNNING", is_type_checking())
  File "/home/sven/x/.venv/src/pythonfinder/src/pythonfinder/environment.py", line 9, in is_type_checking
    from typing import TYPE_CHECKING
ImportError: No module named typing
```

**Comment by techalchemy** [comment]
Ah good point. Does it work now?

**Comment by svenpanne** [comment]
> Ah good point. Does it work now?

I had to install `vistir` from GitHub first (still no compatible) release, but even then Cosmic's "2.7.15+" is still not recognized. The reason is basically the same as without this whole patch: The `version` entry in the dict returned by `parse_python_version` is a `LegacyVersion`, not a `Version`, so it is skipped in the rest of the machinery.

I played around with setting `PYTHONFINDER_IGNORE_UNSUPPORTED=true`, but then one gets:
```
Traceback (most recent call last):
  File "/home/sven/x/.venv/bin/pyfinder", line 11, in <module>
    load_entry_point('pythonfinder', 'console_scripts', 'pyfinder')()
  File "/home/sven/x/.venv/local/lib/python2.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/home/sven/x/.venv/local/lib/python2.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/home/sven/x/.venv/local/lib/python2.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/sven/x/.venv/local/lib/python2.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/home/sven/x/.venv/local/lib/python2.7/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/sven/x/.venv/src/pythonfinder/src/pythonfinder/cli.py", line 33, in cli
    versions = [v for v in finder.find_all_python_versions()]
  File "/home/sven/x/.venv/local/lib/python2.7/site-packages/backports/functools_lru_cache.py", line 137, in wrapper
    result = user_function(*args, **kwds)
  File "/home/sven/x/.venv/src/pythonfinder/src/pythonfinder/pythonfinder.py", line 136, in find_all_python_versions
    major=major, minor=minor, patch=patch, pre=pre, dev=dev, arch=arch, name=name
  File "/home/sven/x/.venv/src/pythonfinder/src/pythonfinder/models/path.py", line 328, in find_all_python_versions
    values = list(self.get_pythons(sub_finder))
  File "/home/sven/x/.venv/src/pythonfinder/src/pythonfinder/models/path.py", line 279, in get_pythons
    reverse=True
  File "/home/sven/x/.venv/src/pythonfinder/src/pythonfinder/models/path.py", line 277, in <genexpr>
    (p for p in self._filter_paths(finder) if p.is_python),
  File "/home/sven/x/.venv/src/pythonfinder/src/pythonfinder/models/path.py", line 264, in <genexpr>
    pth for pth in unnest(finder(p) for p in self.path_entries if p is not None)
  File "/home/sven/x/.venv/src/pythonfinder/src/pythonfinder/utils.py", line 241, in unnest
    for el in target:
  File "/home/sven/x/.venv/src/pythonfinder/src/pythonfinder/models/path.py", line 264, in <genexpr>
    pth for pth in unnest(finder(p) for p in self.path_entries if p is not None)
  File "/home/sven/x/.venv/src/pythonfinder/src/pythonfinder/models/mixins.py", line 76, in find_all_python_versions
    path_filter = filter(None, (sub_finder(p) for p in self.children.values()))
  File "/home/sven/x/.venv/local/lib/python2.7/site-packages/cached_property.py", line 35, in __get__
    value = obj.__dict__[self.func.__name__] = self.func(obj)
  File "/home/sven/x/.venv/src/pythonfinder/src/pythonfinder/models/path.py", line 501, in children
    for child_key, child_val in self._gen_children():
  File "/home/sven/x/.venv/src/pythonfinder/src/pythonfinder/models/path.py", line 493, in _gen_children
    entry = PathEntry.create(path=child, **pass_args)
  File "/home/sven/x/.venv/src/pythonfinder/src/pythonfinder/models/path.py", line 574, in create
    _new = cls(**creation_args)
  File "<attrs generated init f2ce9f44c3bfcd30418aa60bc9df4eef2789996a>", line 16, in __init__
  File "/home/sven/x/.venv/src/pythonfinder/src/pythonfinder/models/path.py", line 523, in get_py_version
    return py_version
UnboundLocalError: local variable 'py_version' referenced before assignment
```

In a nutshell: #46 doesn't work at all for me... :-/

What's wrong with this pull request (#45)? It might not be the prettiest thing, but at least it seems to work.

**Comment by techalchemy** [comment]
What's wrong with it is that it solves for only one very specific case (there is a plus sign on the end of the release), while we are aware of numerous other possible issues which we need to handle -- including conditional logic to fix your specific problem is nice, but it doesn't actually help us with the bigger issue.

**Comment by svenpanne** [comment]
Finally! :smile: The latest version of #46 (4df5c206beb7e9164425e80999e49ccd46e24d1e) works:

```
sven@scotty:~/x$ pipenv run pyfinder --findall
Found python at the following locations:
python2.7: 2.7.15 (None) @ /home/sven/x/.venv/bin/python2.7
python3.6: 3.6.7 (None) @ /usr/bin/python3.6
python3.6m: 3.6.7 (None) @ /usr/bin/python3.6m
python: 2.7.15 (None) @ /usr/bin/python
sven@scotty:~/x$ pipenv run pyfinder --find 2.7.15
Searching for python: 2.7.15
Found python at the following locations:
python: 2.7.15 (None) @ /home/sven/x/.venv/bin/python
```

I guess we should close my pull request #45 in favor of your #46.

**Comment by techalchemy** [comment]
It took some work but I actually wrote some tests to capture this by mocking the output of a subprocess call, a fake path, etc. and asserting that they give a proper `Version` instance. Now I just have a regex that captures versions a bit more forgivingly based on pythons version spec but ignoring anything that isn’t allowed.
